Translating docs/accessibility (#60)
reactjs/de.reactjs.org: content/docs/accessibility.md

}
```

Weitere Infos findest in der [Dokumentation über Fragments](/docs/fragments.html).

## Barrierefreie Formulare {#accessible-forms}

### Etikettierung {#labeling}

> Jedes HTML Formularelement, wie zum Beispiel `<input>` und `<textarea>`, muss barrierefrei etikettiert werden. Beschreibende `label`, welche Für Screenreader relevant sind, müssen vorhanden sein.  

Author: "... muss als barrierefrei gekennzeichnet werden." sounds wrong to me, i would write "muss barrierefrei gekennzeichnet werden."

> Jedes HTML Formularelement, wie zum Beispiel `<input>` und `<textarea>`, muss barrierefrei etikettiert werden. Beschreibende `label`, welche Für Screenreader relevant sind, müssen vorhanden sein.  

Reviewer: Jep :) I updated the suggestion

```javascript{1}
<label htmlFor="namedInput">Name:</label>
<input id="namedInput" type="text" name="name"/>
```

### Den Nutzer über Fehler informieren  {#notifying-the-user-of-errors}

Fehlersituation müssen von allen Nutzern gleich verstanden werden. The foglenden Links zeigen uns wie man Fehlermeldungen auch für Screenreader interpretier gestaltet:

- [Das W3C demonstriert Fehlermeldungen](https://www.w3.org/WAI/tutorials/forms/notifications/)
- [WebAIM nimmt sich der Formularvalidierung an](https://webaim.org/techniques/formvalidation/)

## Fokuskontrolle {#focus-control}

Stelle sicher dass deine Webanwednung vollständig per Tastatur bedienbar ist: 

- [WebAIM sprich über Tastaturbarrierefreiheit](https://webaim.org/techniques/keyboard/)

### Tastaturfokus und die Fokusaußenlinie {#keyboard-focus-and-focus-outline}

Der Tastaturfokus zeigt auf das aktuell ausgewählte DOM-Element welches Eingaben von der Tastatur annimt. Man sieht dies überall anhand der Fokusaußenlinie, genau wie Diese im folgenden Bild:

<img src="../images/docs/keyboard-focus.png" alt="Blue keyboard focus outline around a selected link." />

> Benutze CSS zum Entfernen dieser Linie, zum Beispiel indem du `outline: 0`benutzt, außschließlich nur wenn du Diese auch durch eine andere Implementierung einer Fokusaußenlinie ersetzt. 

Author: I don't think that "focus outline" translates as "Fokussierung".
Maybe something like "Fokus-Markierung"?

> Benutze CSS zum Entfernen dieser Linie, zum Beispiel indem du `outline: 0`benutzt, außschließlich nur wenn du Diese auch durch eine andere Implementierung einer Fokusaußenlinie ersetzt. 

Reviewer: I updated the suggestion

### Mechanismen um direkt zum wichtigsten Inhalt zu springen {#mechanisms-to-skip-to-desired-content}

Stelle einen Mechanismus zur verfügung welcher die Naviagtionselemente deiner Webanwendung überspringt. Dies beschleunigt die Navigation mit der Tastatur.

> Skiplinks, odr auch Skip Navigation Links sind versteckte Navigationslinks die nur dann sichtbar werden wenn Tastaturnutzer mit der Seite interagieren. Diese sind sehr einfach via internen Seitenankern und etwas Styling zu implementieren:

Reviewer: ```suggestion
Skiplinks, oder auch Skip-Navigation-Links sind versteckte Navigations-Links die nur dann sichtbar werden wenn Tastaturnutzer mit der Seite interagieren. Diese sind sehr einfach mit internen Seitenankern und etwas Styling umzusetzen:
```

- [WebAIM - Skip Navigation Links](https://webaim.org/techniques/skipnav/)

Also use landmark elements and roles, such as `<main>` and `<aside>`, to demarcate page regions as assistive technology allow the user to quickly navigate to these sections.
Nutze Außerdem semantische Elemente und Rollen, wie zum Beispiel `<main>`und `<aside>` um verschiedene Bereiche deiner Seite voneinander abzugrenzen, da es entprechende Assitenztechnologien so dem User erlauben sich schnell zwischen diesen Elementen zurecht zu finden.  

> Lies hier mehr über diese für die Barrierefreiheit förderlichen Elemente hier:

Reviewer: ```suggestion
Mehr über den Einsatz dieser Elemente zur Verbesserung der Barrierefreiheit erfährst du hier:
```

- [Accessible Landmarks](https://www.scottohara.me/blog/2018/03/03/landmarks.html)

### Den Fokus programmatisch verwalten {#programmatically-managing-focus}

> Unsere React-Applikation verändern den HTML-DOM kontinuierlich zur Laufzeit, dies führt dazu dass der Tastatur-Fokus gelegentlich verloren geht oder auf ein unvorhersehbares Element gesetzt wird.

Reviewer: ```suggestion
Unsere React-Anwendungen verändern das HTML-DOM kontinuierlich zur Laufzeit, dies führt dazu, dass der Tastatur-Fokus gelegentlich verloren geht oder auf ein unvorhersehbares Element gesetzt wird.
```

> Um diesem Umstand zuvor zu kommen müssen wir hier programmatisch nachhelfen. Zum Beispiel indem wir den Tastatur Fokus weider auf den Button setzen, welcher ein Modal geöffnet hat, nachdem dieses geschlossen wurde.

Reviewer: ```suggestion
Um diesem Umstand zuvor zu kommen, müssen wir hier programmatisch nachhelfen. Zum Beispiel indem wir den Tastatur-Fokus wieder auf den Button setzen, welcher ein Modal geöffnet hat, nachdem dieses geschlossen wurde.
```

> MDN Web Docs takes a look at this and describes how we can build [keyboard-navigable JavaScript widgets](https://developer.mozilla.org/en-US/docs/Web/Accessibility/Keyboard-navigable_JavaScript_widgets).

Reviewer: Missing translation.

Die MDN Web Docs beschreiben wie wir [Tastaturgesteurte JavaScript Widgets](https://developer.mozilla.org/en-US/docs/Web/Accessibility/Keyboard-navigable_JavaScript_widgets) realisieren können.

> Um per React einen Fokus zu setzen, können wir [Refs auf DOM-Elemente](/docs/refs-and-the-dom.html) nutzen.

Reviewer: ```suggestion
Um mit React einen Fokus zu setzen, können wir [Refs auf DOM-Elementen](/docs/refs-and-the-dom.html) nutzen.
```

 }
 ```

> Gelegentlich ist es nötig dass ein Eletern-Component den Fokus auf ein Element eines Kind-Components setzen muss. Dies lässt sich mit dem [Bereitstellen von DOM-Referenzen eines Eletern-Components](/docs/refs-and-the-dom.html#exposing-dom-refs-to-parent-components) via einer speziellen Property, welche die Referenz des Eleternelements zum Kind-Component weitergibt, umsetzen. 

Reviewer: ```suggestion
Gelegentlich ist es nötig, dass eine Eltern-Komponente den Fokus auf ein Element einer Kind-Komponente setzen muss. Dies lässt sich mit dem [Bereitstellen von DOM-Referenzen einer Eltern-Komponente](/docs/refs-and-the-dom.html#exposing-dom-refs-to-parent-components) mit einer speziellen Eigenschaft, welche die Referenz der Eltern-Komponente zur Kind-Komponente weitergibt, umsetzen. 
```

this.inputElement.current.focus();
```

Falls du einen Higher-Order-Component nutzt empfiehtl es sich die [Referenz zum umschlossenen Component weiterzugeben](/docs/forwarding-refs.html) indem du die `forwardRef` React-Funktion nutzt.
Sollte ein Third-Party-HOC diese Technik nicht implentiert haben kannst du das oben verwendete Pattern als einen Fallback nutzen.  

Ein gutes Fokus-Beispielt ist das [react-aria-modal](https://github.com/davidtheclark/react-aria-modal). Dies ist ein relativ rares Beispiel eines vollständig barrierefreiem Modal-Fenster. Hier wird der initiale Fokus auf den Abbrechen Button gesetzt (Um den User davon abzuhalten versehentlich die Success-Action zu aktivieren) und der Fokus kann nicht aus dem Modal-Fenster springen. Nachdem das Modal geschlossen wird wird der Fokus wieder zurück auf das Element gesetzt welches das Modal-Fenster getriggered hat. 

> >Notiz:

Reviewer: ```suggestion
>Hinweis:
```

>

> >Es handelt sich hier um ein wichtiges Feature welches jedoch in Vernünftigen Rahmen genutzt werden sollte. Nutze dieses Feature um den Tastatur-Fokus anzupassen falls der Flow der App gestört sein sollte und versuche nicht die Tastaturnutzung des Users zu antizipieren.

Reviewer 2: ```suggestion
>Es handelt sich hier um ein wichtiges Feature, welches jedoch in einem vernünftigen Rahmen genutzt werden sollte. Nutze dieses Feature, um den Tastatur-Fokus anzupassen, falls der Flow der App gestört sein sollte und versuche nicht, die Tastaturnutzung des Users zu antizipieren.
```

## Maus- und Pointerevents {#mouse-and-pointer-events}

> Stelle sicher dass alle Funktionalitäten welche durch die Maus oder den Zeiger hervorgerufen werden auch via Tastatur gesteuert werden können. Sich komplett von einem Zeigergerät abhängig zu machen führt dazu dass deine App für viele Tastaturnutzer unbrauchbar wird.

Reviewer 2: ```suggestion
Stelle sicher, dass alle Funktionalitäten, welche durch die Maus oder den Zeiger hervorgerufen werden, auch via Tastatur gesteuert werden können. Sich komplett von einem Zeigergerät abhängig zu machen, führt dazu, dass deine App für viele Tastaturnutzer unbrauchbar wird.
```

> Um Dies zu veranschaulichen sehen wir uns einmal ein Anwendungsbeispiel für mangelhafte Barrierefreiheit wegen Click-Events and. Hier siehst du ein Pärifärklick-Beispiel, in welchem der User eine geöffnete Popup-Nachricht schließen kann indem er außerhalb des Elements klickt.

Reviewer 2: ```suggestion
Um dies zu verdeutlichen, schauen wir uns ein Anwendungsbeispiel an, bei dem die Barrierefreiheit durch Click-Events gestört ist. Hier siehst du das "outside-click"-Pattern, in welchem der User eine geöffnete Popup-Nachricht schließen kann, indem er außerhalb des Elements klickt.
```

@ph1p
Hast du hier eine gute Idee?
"outside click pattern" klingt recht komplex.

Reviewer: "Außenklick-Pattern" vielleicht? Es handelt sich ja hier um einen Klick außerhalb eines Popups.

}
```

> Dies würde natürlich mit Zeigergeräten, wie zB. der Maus, super funktionieren, mit der Tastatur alleine führt dies jedoch zu mangelhafter Funktionalität wenn du das nächste Element per tab erreichst. In diesem Fall das `click`-Event nie auf dem `window`-Objekt aufgerufen. Dies kann zu obskuren Fehlern führen, welche es manchen Usern unmöglich macht deine App zu nutzen.

Reviewer: ```suggestion
Dies würde natürlich mit Zeigergeräten, wie zB. der Maus, super funktionieren, mit der Tastatur alleine führt dies jedoch zu mangelhafter Funktionalität wenn du das nächste Element per Tab erreichst. In diesem Fall das `click`-Event nie auf dem `window`-Objekt aufgerufen. Dies kann zu obskuren Fehlern führen, welche es manchen Usern unmöglich macht deine App zu nutzen.
```

<img src="../images/docs/outerclick-with-keyboard.gif" alt="Ein Togglebutton welcher eine Popoverliste öffnet, die nur per Click-Event wieder zu schließen ist." />

> Die selbe Funktionalität kann auch durch angemessene Eventhandler, wie zum Beispiel `onBLur` und `onFocus`, gewährleistet werden:

Reviewer: ```suggestion
Die gleiche Funktionalität kann durch geeignete Eventhandler, wie zum Beispiel `onBLur` und `onFocus`, erreicht werden:
```
